File: extensions/client_matlab/client_stream_rm_vlc.m
%%
% This script receives video from one of the HoloLens sideview grayscale
% cameras and plays it.
% Close the figure to stop.

%% Settings

% HoloLens address
host = '192.168.1.7';

% Port
% Options:
% hl2ss.stream_port.RM_VLC_LEFTFRONT
% hl2ss.stream_port.RM_VLC_LEFTLEFT
% hl2ss.stream_port.RM_VLC_RIGHTFRONT
% hl2ss.stream_port.RM_VLC_RIGHTRIGHT
port = hl2ss.stream_port.RM_VLC_LEFTFRONT;

%%

client = hl2ss.mt.sink_rm_vlc(host, port);
calibration = client.download_calibration();
client.open();

h = []; % figure handle

try
while (true)
    data = client.get_packet_by_index(-1); % -1 for most recent frame
    if (data.status == 0) % got packet
        frame = data.image;
        if (isempty(h))
            h = image(frame);
            colormap gray
        else
            h.CData = frame;
        end
        drawnow
    else % no data
        pause(1); % wait for data
    end
end
catch ME
    disp(ME.message);
end

client.close();

File: viewer/client_stream_rm_vlc.py
#------------------------------------------------------------------------------
# This script receives video from one of the HoloLens sideview grayscale
# cameras and plays it. The camera resolution is 640x480 @ 30 FPS. The stream 
# supports three operating modes: 0) video, 1) video + rig pose, 2) query 
# calibration (single transfer).
# Press esc to stop.
#------------------------------------------------------------------------------

from pynput import keyboard

import cv2
import hl2ss_imshow
import hl2ss
import hl2ss_lnm

# Settings --------------------------------------------------------------------

# HoloLens address
host = "192.168.1.7"

# Port
# Options:
# hl2ss.StreamPort.RM_VLC_LEFTFRONT
# hl2ss.StreamPort.RM_VLC_LEFTLEFT
# hl2ss.StreamPort.RM_VLC_RIGHTFRONT
# hl2ss.StreamPort.RM_VLC_RIGHTRIGHT
port = hl2ss.StreamPort.RM_VLC_LEFTFRONT

# Operating mode
# 0: video
# 1: video + rig pose
# 2: query calibration (single transfer)
mode = hl2ss.StreamMode.MODE_1

# Framerate denominator (must be > 0)
# Effective framerate is framerate / divisor
divisor = 1 

# Video encoding profile and bitrate (None = default)
profile = hl2ss.VideoProfile.H265_MAIN
bitrate = None

#------------------------------------------------------------------------------

if (mode == hl2ss.StreamMode.MODE_2):
    data = hl2ss_lnm.download_calibration_rm_vlc(host, port)
    print('Calibration data')
    print('Image point to unit plane')
    print(data.uv2xy)
    print('Extrinsics')
    print(data.extrinsics)
    print('Undistort map')
    print(data.undistort_map)
    print('Intrinsics (undistorted only)')
    print(data.intrinsics)
    quit()

enable = True

def on_press(key):
    global enable
    enable = key != keyboard.Key.esc
    return enable

listener = keyboard.Listener(on_press=on_press)
listener.start()

client = hl2ss_lnm.rx_rm_vlc(host, port, mode=mode, divisor=divisor, profile=profile)
client.open()

while (enable):
    data = client.get_next_packet()

    print(f'Frame captured at {data.timestamp}')
    print(f'Sensor Ticks: {data.payload.sensor_ticks}')
    print(f'Exposure: {data.payload.exposure}')
    print(f'Gain: {data.payload.gain}')
    print(f'Pose')
    print(data.pose)

    cv2.imshow('Video', data.payload.image)
    cv2.waitKey(1)

client.close()
listener.join()
